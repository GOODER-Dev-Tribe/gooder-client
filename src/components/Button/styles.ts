import styled, { css, DefaultTheme } from 'styled-components'
import { ButtonProps } from '.'

type WrapperProps = Pick<ButtonProps, 'size' | 'hasTooltip' | 'variation'>

const WrapperModifiers = {
  medium: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.small};
  `,
  large: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xxlarge};
  `,
  default: (theme: DefaultTheme) => css`
    background: ${theme.colors.primary};
    color: ${theme.colors.white};
  `,
  withBorder: (theme: DefaultTheme) => css`
    background: none;
    color: ${theme.colors.primary};
    border: 0.1rem solid ${theme.colors.primary};
  `,
  widthIcon: (theme: DefaultTheme) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    height: 4rem;
    width: 4rem;
    border-radius: 100%;
    svg {
      width: 1.5rem;
      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, hasIcon, variation }) => css`
    background: ${theme.colors.primary};
    color: ${theme.colors.white};
    border: 0;
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};
    cursor: pointer;
    ${!!size && WrapperModifiers[size](theme)};
    ${hasIcon && WrapperModifiers.widthIcon(theme)};
    ${!!variation && WrapperModifiers[variation](theme)};
  `}
`
