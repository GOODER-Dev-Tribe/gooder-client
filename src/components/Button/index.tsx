import * as S from './styles'

export type ButtonProps = {
  children?: React.ReactNode
  size?: 'medium' | 'large'
  variation?: 'default' | 'withBorder'
  hasTooltip?: boolean
  icon?: JSX.Element
  onClick?: () => (event: React.MouseEvent<HTMLButtonElement>) => void
}

const Button = ({
  children,
  icon,
  size = 'medium',
  variation = 'default',
  hasTooltip = false,
  ...props
}: ButtonProps) => (
  <S.Wrapper
    size={size}
    variation={variation}
    hasTooltip={hasTooltip}
    hasIcon={!!icon}
    {...props}
  >
    {!!icon && icon}
    {!!children && <span>{children}</span>}
  </S.Wrapper>
)

export default Button
