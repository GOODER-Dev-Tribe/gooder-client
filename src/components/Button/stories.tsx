import { Story, Meta } from '@storybook/react/types-6-0'
import Button from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    }
  }
} as Meta

export const Default: Story = (args) => (
  <div
    style={{ display: 'flex', justifyContent: 'center', alignItems: 'center' }}
  >
    <Button {...args} />
  </div>
)

Default.args = {
  children: 'Continuar'
}
