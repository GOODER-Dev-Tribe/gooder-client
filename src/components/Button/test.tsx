import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'

import Button from '.'

describe('<Button />', () => {
  it('should render a medium button by default', () => {
    const { container } = renderWithTheme(<Button>Continuar</Button>)
    expect(screen.getByRole('button', { name: /Continuar/i })).toHaveStyle({
      height: '3rem',
      padding: '0.8rem 2.4rem',
      'font-size': '1.2rem'
    })
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the large size', () => {
    renderWithTheme(<Button size="large">Buy now</Button>)
    expect(screen.getByRole('button', { name: /Buy now/i })).toHaveStyle({
      height: '4rem',
      padding: '0.8rem 12rem',
      'font-size': '1.6rem'
    })
  })
  it('should render the large variation withBorder', () => {
    renderWithTheme(
      <Button size="large" variation="withBorder">
        Continuar
      </Button>
    )
    expect(screen.getByRole('button', { name: /Continuar/i })).toHaveStyle({
      background: 'none',
      color: '#FF6435',
      border: '0.1rem solid #FF6435'
    })
  })
  it('should render the large variation default', () => {
    renderWithTheme(<Button size="large">Continuar</Button>)
    expect(screen.getByRole('button', { name: /Continuar/i })).toHaveStyle({
      background: '#FF6435',
      color: '#FFF'
    })
  })
})
