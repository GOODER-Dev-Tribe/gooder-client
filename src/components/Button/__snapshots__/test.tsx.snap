// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a medium button by default 1`] = `
.c0 {
  background: #FF6435;
  color: #fff;
  border: 0;
  border-radius: 2rem;
  padding: 0.8rem;
  cursor: pointer;
  height: 3rem;
  font-size: 1.2rem;
  padding: 0.8rem 2.4rem;
  background: #FF6435;
  color: #fff;
}

<button
  class="c0"
>
  <span>
    Continuar
  </span>
</button>
`;
